### PR TITLE
[ENH] Add minimal AptaMCTS pipeline (MCTS-based aptamer recommendation)

### LINKED ISSUE #21: Clean implementation of AptaMCTS
_UPDATED: 07-04-2025_

Add a clean implementation of AptaMCTS, an aptamer recommendation pipeline which combines RandomForest and Monthe Carlo Tree Search (MCTS). These components are the model/classifier and the string optimization algorithm, respectively.

**Todo.** Add a clean, simplified Python implementation of AptaMCTS's model/classifier and preprocessing pipeline to the repository. The MCTS algorithm is already implemented and can easily be used by model with a compatible interface (e.g., see AptaNet and AptaTrans).

High-level steps could be the following:
- Look at the authors' code, specifically the `paper_version` folder, for insights into the original implementation.
- Refactor and add to the repository the model/classifier (in `classifier.py`) and any other data pre-processing logic characteristic of AptaMCTS.
- Implement corresponding test batteries.

AptaMCTS is mostly a RandomForest classifier and leverages `sklearn`. I would expect a very similar interface to that of AptaNet, with the following classes:
- A pipeline module, `aptamcts._pipeline.py`, for the overall pipeline logic. 
- If needed, a model module, `aptamcts._feature_classifier.py`, implementing a classifier and regressor version. This module could also contain specific data preprocessing steps, if any.
- An experiment module, `experiments._aptamer_aptamcts.py`, so that AptaMCTS classifier may be combined with the current implementation of MCTS.

In the end, we want a clean AptaMCTS implementation, easily usable for training and predicting, and easily usable in combination with the current implementation of MCTS for recommending aptamer candidates.

**Code:** [https://github.com/leekh7411/Apta-MCTS](https://github.com/leekh7411/Apta-MCTS).
**Paper:** Lee, Gwangho, et al. "Predicting aptamer sequences that interact with target proteins using an aptamer-protein interaction classifier and a Monte Carlo tree search approach." PloS one 16.6 (2021): e0253760.

### PR DESCRIPTION
**Pull request by kris70lesgo**
#### Reference Issues/PRs

Fixes #21

---

#### What does this implement/fix? Explain your changes.

This PR adds a minimal implementation of **AptaMCTS**, an aptamer recommendation pipeline that combines a scoring model with the existing Monte Carlo Tree Search (MCTS) algorithm.

Key additions:

* **AptaMCTSPipeline**: inference-first pipeline with `predict()` and `recommend()` methods
* **AptamerEvalAptaMCTS**: experiment adapter bridging the pipeline with MCTS via `evaluate()`
* **Minimal feature encoder** in `utils`: placeholder iCTF-style encoding (`pairs_to_features`)
* Integration with existing MCTS without modifying its implementation

The pipeline follows the AptaTrans-style design:

* Uses MCTS through `experiment.evaluate()` only
* Does not require training inside the pipeline
* Assumes a pre-trained model exposing `predict_proba()`

---

#### What should a reviewer concentrate their feedback on?

* Correctness of MCTS integration via experiment adapter
* API consistency with AptaTrans (`predict`, `recommend`)
* Output format of `recommend()` (candidate, sequence, score)
* Placement and minimal design of preprocessing in `utils`
* Overall alignment with repository architecture (no MCTS modifications)

---

#### Did you add any tests for the change?

Yes.

Added focused tests covering:

* `predict()` returns correct type (`np.float64`) and uses `predict_proba`
* `recommend()` output structure and MCTS integration
* `AptamerEvalAptaMCTS.evaluate()` correctness
* Edge cases (invalid inputs, model validation)

All tests pass successfully.

---

#### Any other comments?

* The current feature encoding is a **minimal placeholder**, not a full iCTF implementation, as no reference implementation exists in the repository.
* This is intended to enable integration and can be extended in future work.
* MCTS implementation remains unchanged.

---

#### PR checklist

* [x] The PR title starts with [ENH]
* [x] Added/modified tests
* [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks

### CODE REVIEW

File: pyaptamer/aptamcts/_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+__all__ = ["AptaMCTSPipeline"]
+
+import numpy as np
+
+from pyaptamer.experiments import AptamerEvalAptaMCTS
+from pyaptamer.mcts import MCTS
+from pyaptamer.utils._aptamcts_utils import pairs_to_features
+
+
+class AptaMCTSPipeline:
```
Comment (NennoMP) [inline pyaptamer/aptamcts/_pipeline.py]: Interface is not in line with AptaNet, please look carefully at the description.

File: pyaptamer/experiments/_aptamer_aptamcts.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+__all__ = ["AptamerEvalAptaMCTS"]
+
+import numpy as np
+
+from pyaptamer.experiments._aptamer import BaseAptamerEval
+
+
+class AptamerEvalAptaMCTS(BaseAptamerEval):
```
Comment (NennoMP) [inline pyaptamer/experiments/_aptamer_aptamcts.py:8]: Once the interface is similar to AptaNet as specified in the issue, this could easily be one single class that works for both AptaMCTS and AptaNet, since we expect a very similar interface.

**Review by NennoMP — CHANGES_REQUESTED** [review]
The issue specifically mentions to look at AptaNet interface (quoting myself `I would expect a very similar interface to that of AptaNet`) rather than AptaTrans. Indeed, I don't see methods such as `fit` or `predict_proba`. AptaTrans was mentioned purely in terms of compatibility w.r.t. MCTS.

**Comment by kris70lesgo** [comment]
@NennoMP  Thanks for the review

- I’ve refactored the AptaMCTS pipeline to align with the AptaNet interface:
- Added fit, predict_proba, and predict methods
- Updated the experiment adapter to use predict_proba on (aptamer, target) pairs
- Ensured compatibility with MCTS via experiment.evaluate
- Updated tests accordingly

Would appreciate another review

**Comment by NennoMP** [comment]
I don't see a module in line with what `_feature_classifier.py` does for AptaNet, i.e. defining a classifier and regressor.
